src/lib/utils/package-manifest.ts

```diff
@@ -0,0 +1,193 @@
+ * Returns undefined if packageJSON does not define a Yarn workspace
+ * @param packageJSON the package json object
+ */
+function getPackagePaths(
+    packageJSON: Record<string, unknown>
+): string[] | undefined {
+    if (Array.isArray(packageJSON.workspaces)) {
+        return packageJSON.workspaces;
+    }
+    if (
+        typeof packageJSON.workspaces === "object" &&
+        packageJSON.workspaces != null &&
+        hasOwnProperty(packageJSON.workspaces, "packages") &&
+        Array.isArray(packageJSON.workspaces.packages)
+    ) {
+        return packageJSON.workspaces.packages;
+    }
+    return undefined;
+}
+
+/**
+ * Should produce the same results as the equivalent code in Yarn
+ * https://github.com/yarnpkg/yarn/blob/a4708b29ac74df97bac45365cba4f1d62537ceb7/src/config.js#L799
+ */
+function globPackages(workspacePath: string, packageJsonDir: string): string[] {
+    return glob.sync(resolve(packageJsonDir, workspacePath, "package.json"), {
+        ignore: resolve(packageJsonDir, workspacePath, "node_modules"),
+    });
+}
+
+/**
+ * Given a list of (potentially wildcarded) package paths,
+ * return all the actual package folders found.
+ */
+export function expandPackages(
+    logger: Logger,
+    packageJsonDir: string,
+    workspaces: string[]
+): string[] {
+    // Technnically npm and Yarn workspaces don't support recursive nesting,
+    // however we support the passing of paths to either packages or
+    // to the root of a workspace tree in our params and so we could here
+    // be dealing with either a root or a leaf. So let's do this recursively,
+    // as it actually is simpler from an implementation perspective anyway.
+    return flatMap(workspaces, (workspace) => {
+        const globbedPackageJsonPaths = globPackages(workspace, packageJsonDir);
+        return flatMap(globbedPackageJsonPaths, (packageJsonPath) => {
+            const packageJson = loadPackageManifest(logger, packageJsonPath);
+            if (packageJson === undefined) {
+                logger.error(`Failed to load ${packageJsonPath}`);
+                return [];
+            }
+            const packagePaths = getPackagePaths(packageJson);
+            if (packagePaths === undefined) {
+                // Assume this is a single package repo
+                return [dirname(packageJsonPath)];
+            }
+            // This is a workpace root package, recurse
+            return expandPackages(
+                logger,
+                dirname(packageJsonPath),
+                packagePaths
+            );
+        });
+    });
+}
+
+/**
+ * Finds the corresponding TS file from a transpiled JS file.
+ * The JS must be built with sourcemaps.
+ */
+function getTsSourceFromJsSource(
+    logger: Logger,
+    jsPath: string
+): string | undefined {
+    const contents = readFile(jsPath);
+    const sourceMapPrefix = "//# sourceMappingURL=";
+    const searchResult = contents.search(
```

Comment:
Am I missing some reason `contents.indexOf("\n//# sourceMappingURL=")` doesn't work?

Reply:
Seems like it would work.